csrc/host_ir/jit.cpp

```diff
@@ -0,0 +1,323 @@
+      allocate->buffer()->as<TensorView>()->getLogicalDomain());
+  int64_t ndim = std::ssize(logical_domain);
+  llvm::Value* ndim_val = llvm::ConstantInt::get(int64_type, ndim);
+  // Check shape ndim
+  dimCheck(shape_ndim_arg, ndim_val, builder);
+  // Check strides ndim
+  dimCheck(strides_ndim_arg, shape_ndim_arg, builder);
+
+  // Create constants for type and device from params
+  at::ScalarType data_type = data_type_to_aten(
+      allocate->buffer()->dtype() == DataType::Index
+          ? PrimDataType::Int
+          : allocate->buffer()->dtype());
+  llvm::Value* dtype_constant =
+      llvm::ConstantInt::get(int32_type, static_cast<int32_t>(data_type));
+  llvm::Value* device_index_constant = llvm::ConstantInt::get(
+      int64_type, Communicator::getInstance().deviceId());
+
+  // Get the at::native::empty_strided_cuda function pointer (registered in the
+  // JIT)
+  llvm::Function* at_empty_strided_cuda_func =
+      mod->getFunction(kHostIrJitAtEmptyStridedCudaFuncName);
+  if (at_empty_strided_cuda_func == nullptr) {
+    llvm::FunctionType* at_empty_strided_cuda_func_type =
+        llvm::FunctionType::get(
+            builder.getVoidTy(),
+            {int64_ptr_type,
+             int64_type,
+             int64_ptr_type,
+             int64_type,
+             int32_type,
+             int64_type,
+             void_ptr_type},
+            false);
+    at_empty_strided_cuda_func = llvm::Function::Create(
+        at_empty_strided_cuda_func_type,
+        llvm::Function::ExternalLinkage,
+        kHostIrJitAtEmptyStridedCudaFuncName,
+        mod);
+  }
+
+  // Call at::native::empty_strided_cuda with constants
+  builder.CreateCall(
+      at_empty_strided_cuda_func,
+      {sizes_arg,
+       shape_ndim_arg,
+       strides_arg,
+       strides_ndim_arg,
+       dtype_constant,
+       device_index_constant,
+       out_tensor_arg});
+
+  builder.CreateRetVoid();
+
+  // Verify the module
+  std::string error;
+  llvm::raw_string_ostream error_stream(error);
+  NVF_ERROR(
+      !llvm::verifyModule(*mod, &error_stream),
+      "LLVM module verification failed: " + error);
+
+  const bool debug_print = isDebugDumpEnabled(DebugDumpOption::HostIrJit);
+  if (debug_print) {
+    llvm::outs() << "=== LLVM IR ===\n";
+    mod->print(llvm::outs(), nullptr);
+  }
+}
+
+void compile(
+    const hir::HostIrContainer* container,
+    llvm::orc::LLJIT* jit,
+    std::unordered_map<const kir::Allocate*, AllocateFunc>& allocate_funcs) {
+  FUSER_PERF_SCOPE("HostIrJit::compile");
+  NVF_ERROR(
+      container != nullptr,
+      "container is nullptr during host ir JIT compilation");
+  auto ctx = std::make_unique<llvm::LLVMContext>();
+  auto mod = std::make_unique<llvm::Module>("host_ir_jit_module", *ctx);
+  std::vector<kir::Allocate*> allocate_exprs;
+  for (auto* expr : container->topLevelExprs()) {
+    if (auto* allocate = dynamic_cast<kir::Allocate*>(expr)) {
+      createAndInsertAllocationFunction(allocate, mod.get());
```

Comment:
Is it possible to let createAndInsertAllocationFunction return the function address so we can add it directly to `allocate_funcs`? This way, we don't need to add function addresses in [a separate loop](https://github.com/NVIDIA/Fuser/pull/4651/files#diff-c0568f66b12e6643f5dc2d89e0c64cfe6e3585f8c7ad3f99b95ef8b7fadddabcR234-R241).